ffi/Gosu.cpp

```diff
@@ -1,251 +1,200 @@
-#include <Gosu/Gosu.hpp>
+#include "Gosu_FFI_internal.h"
 
-extern "C" {
-#include "Gosu.h"
-
-void Gosu_gl_z(double z, void function(void *data), void *data)
+GOSU_FFI_API void Gosu_gl_z(double z, void function(void*), void* data)
 {
-    std::function<void ()> callback;
-    callback = [=]() { function(data); };
-
-    Gosu::Graphics::gl(z, callback);
+    Gosu::Graphics::gl(z, [=]() { function(data); });
 }
 
-void Gosu_gl(void function(void *data), void *data)
+GOSU_FFI_API void Gosu_gl(void function(void*), void* data)
 {
-    std::function<void ()> callback;
-    callback = [=]() { function(data); };
-
-    Gosu::Graphics::gl(callback);
+    Gosu::Graphics::gl([=] { function(data); });
 }
 
-Gosu_Image *Gosu_render(int width, int height, void function(void *data), void *data, unsigned image_flags)
+GOSU_FFI_API Gosu_Image* Gosu_render(int width, int height, void function(void*), void* data,
+                                     unsigned image_flags)
 {
-    std::function<void ()> callback;
-    callback = [=]() { function(data); };
-
-    return reinterpret_cast<Gosu_Image *>(new Gosu::Image(Gosu::Graphics::render(width, height, callback, image_flags)));
+    Gosu::Image image = Gosu::Graphics::render(width, height, [=] { function(data); }, image_flags);
+    return new Gosu_Image{image};
 }
 
-Gosu_Image *Gosu_record(int width, int height, void function(void *data), void *data)
+GOSU_FFI_API Gosu_Image* Gosu_record(int width, int height, void function(void*), void* data)
 {
-    std::function<void ()> callback;
-    callback = [=]() { function(data); };
-    return reinterpret_cast<Gosu_Image *>(new Gosu::Image(Gosu::Graphics::record(width, height, callback)));
+    Gosu::Image image = Gosu::Graphics::record(width, height, [=] { function(data); });
+    return new Gosu_Image{image};
 }
 
-void Gosu_flush()
+GOSU_FFI_API void Gosu_flush(void)
```

Comment:
@cyberarm `Gosu_flush()` declares a vararg function in C (same as `Gosu_flush(...)`), but we explicitly want zero arguments. Hm, technically I could have left out the `void` here because this is C++, but it needs to be there in the C headers.